Add test cases for antispam rules, fix bug in attachments rule & test

kwzrd opened the pull request:
This PR introduces a series of unit test cases for the bot's antispam rules. This addresses the following issues.

Closes #595 via 1f17d0e.
Closes #596 via b0713be.
Closes #597 via f15f318.
Closes #598 via 4bfe30d.
Closes #603 via f32ffaf.

I also sneak in a quick bugfix in 4199b12: it appears that the already existing rule for attachments contains a slight bug where in the printout, the configured `'max'` key is looked up instead of `'interval'`. The same commit also fixes the key in the rule's unit test, which was incorrectly expecting the wrong key.

Finally, aedbd6f is a refactor of the name `msg` into `make_msg` which prevents a harmless, but unnecessary name collision in the `msg for msg in ...` generator expressions down the module.

Edit:
Closes #599 via 54e8c98.
Closes #602 via cae7f25.

kwzrd commented:
The committed code is considered ready for review. I'm currently awaiting a response from a user in #599 - if they're no longer interested, I'd like to include that issue in this PR's scope as well.

kwzrd commented on tests/bot/rules/test_burst.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def make_msg(author: str) -> MockMessage:
+    """Init a MockMessage instance with author set to `author`.
```
Thanks, addressed in 8456331.

kwzrd commented:
The test cases are similar, yes - the rules are too. What do you propose can be done to make this better? Would it make sense to implement some kind of an "executor" on the side, and then just feed it the test cases from each module?

MarkKoz commented:
Two helper functions: one for "allows" tests and one for "disallows". They should be passed the `apply` function and the subtest arguments. The helpers will do the loop, calling of `apply`, and the assertions.

As I mentioned, some of the subtest arguments follow a similar pattern across test suites. To facilitate this, a function could be defined which will take a `make_msg` function and return subtest arguments. This function can shared among some test suites to generate subtest arguments while other suites with more divergent subtest can still use their own subtest arguments. My only concern is that such function adds a layer of indirection which has the consequence of making unclear what the actual subtest values are.

I believe such a design will reduce redundancy but still remain flexible should test suites need to diverge in code in the future.

kwzrd commented:
93d19f3 adds an ABC (9758e32 actually adds the `metaclass=ABCMeta`), b89f9c5 then adjusts the existing rules to use the ABC. I decide to make the `DisallowedCase` (previously `Case`) more generic by having an `n_violations` attr instead of a rule-specific one. This way, the `subTest` kwargs can be generic as well.

Let me know how this looks, and thanks for the feedback.

MarkKoz commented on tests/bot/rules/__init__.py:
```diff
@@ -0,0 +1,76 @@
+    """
+
+    apply: Callable  # The tested rule's apply function
+    config: Dict[str, int]
```
These _could_ be enforced

```py
    @property
    @abc.abstractmethod
    def apply(self) -> Callable:
        """Docstring"""
        raise NotImplementedError
```

kwzrd replied:
I briefly considered this, but eventually decided not to. Should the child unittest not provide an `apply` attr, the tests will fail with:

```python
AttributeError: 'MyRuleTest' object has no attribute 'apply'
```

I believe that's descriptive enough. I'll mark this as resolved then if that's ok.

kwzrd commented:
I've added tests for `duplicates` (54e8c98) and `newlines` (cae7f25).

As the latter is more complex than the other rules, yielding two distinct error reports, I decided to implement two subclasses of the `RuleTest` ABC for it. Each checks one type of violation.

If the given messages violate both **total count** and **max consecutive**, the rule will fail on the **total** as that is checked first. Since I consider this an implementation detail and not "desired functionality", I do not test it, and each disallowed test case only violates the rule one way or the other, so that the tests still pass should the two checks switch order.

This PR now provides 100% coverage for the entire `rules` package - I've updated the PR description to mention the two extra issues it closes.
